File: solutions/c/secret-handshake/1/secret_handshake.c
#include "secret_handshake.h"

char *actions[] = {
    "wink",
    "double blink",
    "close your eyes",
    "jump"
};

const char **commands(size_t number){
    char **result;
    result = calloc(4, sizeof(char*));
    if(!result) return NULL;
    
    int count = 0;
    for(int i = 0; i < 4; i++){
        if((number >> i) & 1){
            result[count] = malloc(MAX_COMMAND_LEN);
            if(!result[count]){
                for (int j = 0; j < count; j++) free(result[j]);
                free(result);
                return NULL;
            }
            strncpy(result[count], actions[i], MAX_COMMAND_LEN);
            count++;
        }
    }
    // Reverse order
    if((number >> 4) & 1){
        for(int i = 0; i < count/2; i++){
            char *tmp = result[i];
            result[i] = result[count - i - 1];
            result[count - i - 1] = tmp;
        }
    }
    
    return (const char**) result;
}

File: solutions/zig/secret-handshake/1/secret_handshake.zig
const std = @import("std");
const mem = std.mem;

pub const Signal = enum {
    wink,
    double_blink,
    close_your_eyes,
    jump,
};

pub fn calculateHandshake(allocator: mem.Allocator, number: u5) ![]const Signal {
    var result = try std.ArrayList(Signal).initCapacity(allocator, 1);
    errdefer result.deinit(allocator);

    var n: u5 = number;

    if (n & 1 == 1) {
        try result.append(allocator, .wink);
    }
    n >>= 1;
    if (n & 1 == 1) {
        try result.append(allocator, .double_blink);
    }
    n >>= 1;
    if (n & 1 == 1) {
        try result.append(allocator, .close_your_eyes);
    }
    n >>= 1;
    if (n & 1 == 1) {
        try result.append(allocator, .jump);
    }
    n >>= 1;
    // Reverse order
    if (n & 1 == 1) {
        std.mem.reverse(Signal, result.items);
    }

    return try result.toOwnedSlice(allocator);
}
